### ISSUE TITLE
Supers are not lazily evaluated

### ISSUE THREAD

**Issue by Pacman99** [divnix/POP#11]
```
nix-repl> pop.lib.pop { supers = [ { a = throw "err"; b = 5; } ]; }
error: err
```
An attribute in a pop containing an error shouldn't error out the entire pop. Other attributes should still be accessible even if one attribute errors out.

The following code should print out `5`, but it would just error out now.

```
p = pop.lib.pop { supers = [ { a = throw "err"; b = 5; } ]; }
p.b
```

cc @fare

**Comment by fare** [comment]
That sounds like a bug indeed: the list of supers should be eagerly evaluated to resolve the multiple inheritance hierarchy, but the fields of the super itself shouldn't. Maybe the bug though isn't in the super handling, but in the casting of prototype-less record to pop prototype? Even then that would be a bug, but a different one.

**Comment by Pacman99** [comment]
> Maybe the bug though isn't in the super handling, but in the casting of prototype-less record to pop prototype?

I dont think so, I just used the plain attribute set as an example, but it happens even if that attrset was wrapped in a `kPop`.

```nix
pop.lib.pop { supers = [ (kPop { a = throw "err"; b = 5; }) ]; }
```
 Also errors out.

**Comment by GTrunSec** [comment]
@fare  Are there any inquiries pertaining to this matter?  I also got the same problem when I trying embed POP into my instance. Thanks

**Comment by Pacman99** [comment]
I opened this https://github.com/NixOS/nix/issues/8364

**Comment by fare** [comment]
https://twitter.com/Ngnghm/status/1717684526720094339

Lacking the ability to compare structures or functions by pointer identity, or system-provided nonces as identity tags, the only reliable solution is user-provided labels with some naming conventions to retrieve some modularity.

**Comment by fare** [comment]
Akshully, we might be able to do proper pointer equality: https://code.tvl.fyi/about/tvix/docs/value-pointer-equality.md

**Comment by Pacman99** [comment]
Yeah I just saw that thread! I might give that a try to see if we can fix this issue. I tried something similar with lists which sometimes worked, so hopefully this trick will work fully.

**Comment by fare** [comment]
Note that one part of the problem is ensuring that `true` is returned when the two are pointer-equal, the other part of the problem is ensuring that `false` is returned when they are not, instead of an error being thrown or an infinite branch taken.

**Comment by fare** [comment]
Here's a way to "purely" create tags you can compare for equality:

let generateTag = name: {id=_:1; inherit name;} ; t1 = generateTag 1 ; t2 = generateTag 2 ; t2_ = generateTag 2 ; in [(t1 == t1) (t1 == t2) (t2 == t2) (t2 == t2_)]

Or we can compare functions with {fun=next;} == {
The name is memoized purely for debugging reasons, but creating two tags with the same name yields different tags.

Beware that automatically wrapping 

PS: The URL for the document above has changed to https://code.tvl.fyi/tree/tvix/docs/src/value-pointer-equality.md

**Comment by fare** [comment]
@Pacman99 Can you check whether my commit b6b81b1c42ba78fda5b3021f8927cf2a52b5460e fixes the issue?

**Comment by fare** [comment]
Also, I did some renaming, in line with my latest (still unfinished) sequel to the POOF article, my "Lambda, the Ultimate Object" essay.

Now distinguishing specification (the things you instantiate) from target (the thing instantiated), and calling prototype a target that remembers its specification, etc.

**Comment by rebmit** [comment]
Seems like `builtins.elem` here should also use pointer equality, e.g. `builtins.any (t: getId t == getId c)`.

https://github.com/divnix/POP/blob/b6b81b1c42ba78fda5b3021f8927cf2a52b5460e/POP.nix#L166

**Comment by fare** [comment]
The problem is that if it falls back to infinitely deep field comparison when the identity comparison fails, then it may fall into non-termination.
